File: demo-task-executor-java/src/main/java/demo/logger/FileLogger.java
package demo.logger;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Objects;

/**
 * Logger that writes messages to a file.
 * This class IS thread-safe.
 */
public class FileLogger extends AbstractLogger {

    private final File file;

    /**
     * Creates a file logger and deletes its previous content.
     *
     * @param fileName the file logger name.
     */
    public FileLogger(String fileName) {
        this.file = new File(Objects.requireNonNull(fileName));
        file.delete();
    }

    @Override
    synchronized public void log(Severity severity, Object message) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(file, true))) {
            writer.write(getMessagePrefix(severity) + message + "\n");
            writer.flush();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: demo-task-executor-scala/src/main/scala/demo/logger/FileLogger.scala
package demo.logger

import java.io.{BufferedWriter, File, FileWriter}
import java.util.Objects

import demo.AutoCloseableHelper._

/**
  * Logger that writes messages to a file.
  * This class IS thread-safe.
  */
class FileLogger(fileName: String) extends AbstractLogger {

    private val file = new File(Objects.requireNonNull(fileName))
    file.delete()

    override def log(severity: Severity, message: Any) {
        this.synchronized {
            use(new BufferedWriter(new FileWriter(file, true))) {
                writer =>
                    writer.write(getMessagePrefix(severity) + message + "\n")
                    writer.flush()
            }
        }
    }
}
